### PR TITLE
HADOOP-18740. S3A prefetch cache blocks should be accessed by RW locks

### PR DESCRIPTION
**Pull request by virajjasani**
Jira: HADOOP-18740

### CODE REVIEW

**Comment by virajjasani** [comment]
`us-west-2`

```
mvn clean verify -Dparallel-tests -DtestsThreadCount=8 -Dscale -Dprefetch
```

```
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ITestS3ABlockOutputArray.testDiskBlockCreate:114 » IO File name too long
[ERROR]   ITestS3ABlockOutputByteBuffer>ITestS3ABlockOutputArray.testDiskBlockCreate:114 » IO
[ERROR]   ITestS3ABlockOutputDisk>ITestS3ABlockOutputArray.testDiskBlockCreate:114 » IO ...
[ERROR]   ITestS3AFileContextMainOperations>FileContextMainOperationsBaseTest.testCreateFlagCreateAppendNonExistingFile:831 » IO
[ERROR]   ITestS3AFileContextURI>FileContextURIBase.testCreateFileInNonExistingDirectory:164 » IO
[ERROR]   ITestS3AFileContextURI>FileContextURIBase.testDeleteNonExistingFileInDir:357 » IO
[INFO] 
[ERROR] Tests run: 1165, Failures: 0, Errors: 6, Skipped: 148

```

All failures related to HADOOP-18744

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 53s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  35m 41s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 33s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |  15m 42s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m  5s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 29s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 10s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 41s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 38s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 38s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 51s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 42s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |  16m 42s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  15m 41s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |  15m 41s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  1s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 28s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  0s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 41s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 35s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m  5s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 19s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 54s |  |  The patch does not generate ASF License warnings.  |
|  |   | 187m 53s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5675 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux ab2194f686fd 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 912403c4bfd2b5aed1f1a7d94c39cc4b53d44c85 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/1/testReport/ |
| Max. process+thread count | 1235 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by MayankSinghParmar** [comment]
I want to implement a new cache replacement algorithm in hdfs but I don't know which part of module to use and modify. Can any one please help me find the modules and functions which need to override to add my own caching replacement algorithm in hdfs

**Comment by steveloughran** [comment]
@MayankSinghParmar get on the hadoop hdfs mailing list and discuss there.

@virajjasani can you rebase and retest?

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java [hunk 1]
```diff
@@ -99,6 +106,32 @@ public String toString() {
...
     }
+
+    /**
+     * Take the read or write lock.
+     *
+     * @param lockType type of the lock.
+     */
+    void takeLock(LockType lockType) {
+      if (LockType.READ == lockType) {
+        this.lock.readLock().lock();
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: remove the `this.`

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java [hunk 2]
```diff
@@ -99,6 +106,32 @@ public String toString() {
...
+    }
+
+    /**
+     * Release the read or write lock.
+     *
+     * @param lockType type of the lock.
+     */
+    void releaseLock(LockType lockType) {
+      if (LockType.READ == lockType) {
+        this.lock.readLock().unlock();
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: remove the `this.`

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java [hunk 3]
```diff
@@ -268,12 +310,15 @@ public void close() throws IOException {
     int numFilesDeleted = 0;
 
     for (Entry entry : blocks.values()) {
+      entry.takeLock(Entry.LockType.WRITE);
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: should we be acquiring locks in close()?
good: no race condition in close
bad) the usual

also, L303: should closed be atomic?

**Review by steveloughran — CHANGES_REQUESTED** [review]
core changes look good, just a nit about using this. where it isn't needed.

I'm worried about concurrency of close(), where
* we should make the `closed` flag atomic boolean; only execute close on the first call and have all the other ops fail if closed
* be confident that locking in close() is good. After the disaster with the abfs prefetcher, we need to look carefully here

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java (hunk 3 above)
Comment (virajjasani) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: > also, L303: should closed be atomic?

+1 to this suggestion, let me create a separate patch with HADOOP-18756 to better track it.

> good: no race condition in close
> bad) the usual

sounds reasonable, let me try setting timeout

**Comment by virajjasani** [comment]
`us-west-2`:

```
mvn clean verify -Dparallel-tests -DtestsThreadCount=8 -Dscale

mvn clean verify -Dparallel-tests -DtestsThreadCount=8 -Dscale -Dprefetch
```

Test results are clean

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 33s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  34m  1s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  15m 37s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |  14m 21s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m 13s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 35s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 17s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 52s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 41s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  22m  8s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 49s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 52s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |  14m 52s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 19s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |  14m 19s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  6s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 33s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  5s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 51s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 34s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  22m 23s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 32s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  2s |  |  The patch does not generate ASF License warnings.  |
|  |   | 174m 48s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5675 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 9432f0056538 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 710441f19ca08701a4e0f5c32fa77f3ff7ce9736 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/2/testReport/ |
| Max. process+thread count | 1783 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java [hunk 4]
```diff
@@ -310,7 +333,12 @@ public void close() throws IOException {
     int numFilesDeleted = 0;
 
     for (Entry entry : blocks.values()) {
-      entry.takeLock(Entry.LockType.WRITE);
+      boolean lockAcquired = entry.takeLock(Entry.LockType.WRITE, 5, TimeUnit.SECONDS);
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: pull the number into a constant. Know that I automatically -1 all inline constants in production code and save time all round.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java [hunk 5]
```diff
@@ -127,11 +128,33 @@ void takeLock(LockType lockType) {
...
+
+    /**
+     * Try to take the read or write lock within the given timeout.
+     *
+     * @param lockType type of the lock.
+     * @param timeout the time to wait for the given lock.
+     * @param unit the time unit of the timeout argument.
+     * @return true if the lock of the given lock type was acquired.
+     */
+    boolean takeLock(LockType lockType, long timeout, TimeUnit unit) {
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: should this and the others be private? you don't want other classes playing with your lock code...

**Review by steveloughran — CHANGES_REQUESTED** [review]

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java (hunk 4 above)
Comment (virajjasani) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: my bad, let me fix this real quick

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java (hunk 5 above)
Comment (virajjasani) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/prefetch/SingleFilePerBlockCache.java]: sounds good, done

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 49s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  37m 27s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 27s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |  15m 47s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m  6s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 30s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 42s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 42s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 22s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 51s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 42s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |  16m 42s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  15m 45s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |  15m 45s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  0s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 31s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  0s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 42s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  spotbugs  |   2m 33s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m 50s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 22s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 53s |  |  The patch does not generate ASF License warnings.  |
|  |   | 189m 16s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5675 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux ce03b807e480 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d2b9210104b1ec0e4a09f22fcfc6df7b82557017 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/3/testReport/ |
| Max. process+thread count | 1244 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5675/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by steveloughran — APPROVED** [review]
+1. concurrency stuff always worries me, but this looks good

**Pull request merged**